app/controllers/api/base_controller/renderer.rb

```diff
@@ -540,6 +546,61 @@ def render_options(resource, data = {})
         klass = collection_class(resource)
         render :json => OptionsSerializer.new(klass, data).serialize
       end
+
+      # This is a helper method used by both .determine_include_for_find and
+      # .determine_extra_cols to collect and filter virtual_attributes for the
+      # :include_for_find and :extra_cols options that are passed to Rbac. The
+      # intent is to reduce a large amount of shared code between those two
+      # shared methods by combining them into this one.
+      #
+      # The required block used by each of aforementioned methods is used to do
+      # custom filtering that pertains to each of those methods.
+      #
+      def virtual_attributes_for(klass)
+        return nil unless klass.respond_to?(:reflect_on_association)
+
+        type    = @req.subject
+        results = []
+
+        validate_attr_selection(klass).last.each do |vattr|
+          next if vattr == "href_slug"
+
+          attr_name, attr_base = split_virtual_attribute(vattr)
+          filtered_attr        = yield type, attr_name, attr_base
+
+          results << filtered_attr if filtered_attr
+        end
+
+        results.empty? ? nil : results
+      end
+
+      def determine_include_for_find(klass)
+        attrs = virtual_attributes_for(klass) do |type, attr_name, attr_base|
+          next if attr_base.blank?
+          next if virtual_attribute_accessor(type, attr_name)
+          next if Rbac::Filterer::CLASSES_THAT_PARTICIPATE_IN_RBAC.include?(attr_base)
+
+          attr_base
+        end
+
+        # Handle nested relationships and convert to a hash
+        if attrs
+          attrs.each_with_object({}) do |key, include_for_find|
+            nested = include_for_find
+            key.split(".").each { |k| nested = nested[k] ||= {} }
```

Comment:
commented on the other side that I think you want to return nested here
(but you probably already commented over there that this is right... )

Reply:
Yes I did, it is here:

https://github.com/ManageIQ/manageiq-api/pull/887#discussion_r486694177

Reply:
sorry - this really doesn't make sense to me.

for `"host.hardware"` I read:

```ruby
nested = {host: {hardware: {}}}
return [:host, :hardware] # <===
```

Reply:
@kbrock "`nested`" is not the return value, it is a temp value for a single iteration of the `.each_with_object`, and a "tracker variable" for the nested `.each`.  `include_for_find` is the return value of this method.

`nested` is keeping track of the location as we traverse through `include_for_find` as we split the `key` for each iteration.  Because Ruby is "pass by reference", we are manipulating the value of `include_for_find` by passing a reference to the "nested" object each time we traverse deeper into the hash.

This means that for something like `"hardware.host.name"`, the loop for the `.each` (inner loop) would be:

### 1.

```ruby
key = "hardware"
include_for_find_before = {}
include_for_find_after  = {"hardware" => {}}
nested_before           = {} # the value of `include_for_find_before`
nested_after            = {} # the value of `include_for_find_after["hardware"]
```

### 2.
```ruby
key = "host"
include_for_find_before = {"hardware" => {}}
include_for_find_after  = {"hardware" => {"host" => {}}}
nested_before           = {"hardware" => {}} # the value of `include_for_find_before`
nested_after            = {} # the value of `include_for_find_after["hardware"]["host"]`
```

This is also why I wrote a test for this...

> "with nested indirect virtual attribute ('operating_system.computer_system.created_at')"